src/protocol/basics/mul/sparse.rs

```diff
@@ -0,0 +1,428 @@
+    #[test]
+    #[ignore]
+    fn print_mappings() {
+        for &a in ZeroPositions::all() {
+            let a_flags = <[bool; 3]>::from(a);
+            for &b in ZeroPositions::all() {
+                let b_flags = <[bool; 3]>::from(b);
+                println!(
+                    "(Self::{a:?}, Self::{b:?}) => {:?},",
+                    calculate_work(Role::H1, a_flags, b_flags)
+                );
+            }
+        }
+    }
+
+    fn check_punctured_output<F, T>(v: &[T; 3], work: MultiplyZeroPositions)
+    where
+        F: Field,
+        T: Borrow<Replicated<F>>,
+    {
+        for (&role, expect_zero) in zip(Role::all(), <[bool; 3]>::from(work.output())) {
+            if expect_zero {
+                assert_eq!(F::ZERO, v[role as usize].borrow().left());
+                assert_eq!(F::ZERO, v[role.peer(Left) as usize].borrow().right());
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn check_output() {
+        let world = TestWorld::new(QueryId);
+        let mut rng = thread_rng();
+
+        for &a in ZeroPositions::all() {
+            for &b in ZeroPositions::all() {
+                if ZeroPositions::is_pointless((a, b)) {
+                    continue;
+                }
+
+                let v1 = SparseField::new(rng.gen::<Fp31>(), a);
+                let v2 = SparseField::new(rng.gen::<Fp31>(), b);
+                let result = world
+                    .semi_honest((v1, v2), |ctx, (v_a, v_b)| async move {
+                        ctx.multiply_sparse(RecordId::from(0), &v_a, &v_b, (a, b))
+                            .await
+                            .unwrap()
+                    })
+                    .await;
+                check_punctured_output(&result, (a, b));
+                assert_eq!(v1.value() * v2.value(), result.reconstruct());
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn check_output_malicious() {
+        let world = TestWorld::new(QueryId);
+        let mut rng = thread_rng();
+
+        for &a in ZeroPositions::all() {
+            for &b in ZeroPositions::all() {
+                if ZeroPositions::is_pointless((a, b)) {
+                    continue;
+                }
+
+                let v1 = SparseField::new(rng.gen::<Fp31>(), a);
+                let v2 = SparseField::new(rng.gen::<Fp31>(), b);
+                let result = world
+                    .semi_honest((v1, v2), |ctx, (v_a, v_b)| async move {
+                        let v = MaliciousValidator::new(ctx);
+                        let m_ctx = v.context();
+                        let m_a = m_ctx
+                            .upgrade_with_sparse(&BitOpStep::from(0), RecordId::from(0), v_a, a)
+                            .await
+                            .unwrap();
+                        let m_b = m_ctx
+                            .upgrade_with_sparse(&BitOpStep::from(1), RecordId::from(0), v_b, b)
+                            .await
+                            .unwrap();
```

Comment:
Do you think there's any value in doing these in parallel with a `try_join`? Any kind of infra bug that this could pick up?